Write logs to fifo from a driver run as non-root user

### Nomad version
Nomad v1.2.4 (55e5c49b99a6fd2bf925e7fd98d95829776c331f)

### Operating system and Environment details
Void Linux

### Issue
While trying to setup the [podman driver](https://github.com/hashicorp/nomad-driver-podman/) with a non-root user all containers fail to start due to the podman driver not being able to write on the log fifo files. The nomad client is running as root.

This happens due to both the `<alloc-dir>` and the internal `<alloc-dir>/<alloc-id>/alloc/log/.<task>.stdout.fifo` being owned by root and having only user permissions. Even if I grant the podman user the permissions to access to the `<alloc-dir>`, the fifos are being created with the [limited permissions](https://github.com/hashicorp/nomad/blob/main/client/lib/fifo/fifo_unix.go#L21).

What I can get from the [driver documentation](https://www.nomadproject.io/docs/internals/plugins/task-drivers#logging) is that drivers are expected to be able to write to the alloc logs fifo files but because of the permission design it looks like it's not possible to run any driver with a non-root user.

This works well when running podman as root.

### Reproduction steps
Start the podman service as a non-root user and try to run any job

#### Expected Result
Containers starting when running with a podman service with a non-root user

#### Actual Result
Job fails to start due to podman not being able to write to log's fifo

### Job file (if appropriate)
https://github.com/hashicorp/nomad-driver-podman/blob/main/examples/jobs/redis_deprecated.nomad

### Nomad Client logs (if appropriate)

```
$ nomad agent -dev -plugin-dir /usr/lib/nomad/plugins/ -bind 0.0.0.0
    2022-01-29T09:20:09.519Z [DEBUG] worker: dequeued evaluation: worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 eval_id=7c921fcf-641d-9a96-e2f1-7d955cc2fb72 type=service namespace=default job_id=redis node_id="" triggered_by=job-register
    2022-01-29T09:20:09.520Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=7c921fcf-641d-9a96-e2f1-7d955cc2fb72 job_id=redis namespace=default worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3
  results=
  | Total changes: (place 1) (destructive 0) (inplace 0) (stop 0)
  | Created Deployment: "b724ae9c-c092-77dc-06d8-18576acccc8d"
  | Desired Changes for "cache": (place 1) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 0) (canary 0)

    2022-01-29T09:20:09.521Z [DEBUG] http: request complete: method=PUT path=/v1/jobs duration=15.623236ms
    2022-01-29T09:20:09.532Z [DEBUG] worker: submitted plan for evaluation: worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 eval_id=7c921fcf-641d-9a96-e2f1-7d955cc2fb72
    2022-01-29T09:20:09.533Z [DEBUG] worker.service_sched: setting eval status: eval_id=7c921fcf-641d-9a96-e2f1-7d955cc2fb72 job_id=redis namespace=default worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 status=complete
    2022-01-29T09:20:09.534Z [DEBUG] http: request complete: method=GET path=/v1/evaluation/7c921fcf-641d-9a96-e2f1-7d955cc2fb72 duration=3.079759ms
    2022-01-29T09:20:09.537Z [DEBUG] client: updated allocations: index=11 total=1 pulled=1 filtered=0
    2022-01-29T09:20:09.538Z [DEBUG] client: allocation updates: added=1 removed=0 updated=0 ignored=0
    2022-01-29T09:20:09.538Z [DEBUG] worker: updated evaluation: worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 eval="<Eval \"7c921fcf-641d-9a96-e2f1-7d955cc2fb72\" JobID: \"redis\" Namespace: \"default\">"
    2022-01-29T09:20:09.543Z [DEBUG] worker: ack evaluation: worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 eval_id=7c921fcf-641d-9a96-e2f1-7d955cc2fb72 type=service namespace=default job_id=redis node_id="" triggered_by=job-register
    2022-01-29T09:20:09.560Z [DEBUG] http: request complete: method=GET path=/v1/evaluation/7c921fcf-641d-9a96-e2f1-7d955cc2fb72/allocations duration=2.834149ms
    2022-01-29T09:20:09.572Z [DEBUG] client: allocation updates applied: added=1 removed=0 updated=0 ignored=0 errors=0
    2022-01-29T09:20:09.576Z [DEBUG] client.alloc_runner.task_runner: lifecycle start condition has been met, proceeding: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis
    2022-01-29T09:20:09.579Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: starting plugin: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis path=/usr/bin/nomad args=["/usr/bin/nomad", "logmon"]
    2022-01-29T09:20:09.580Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: plugin started: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis path=/usr/bin/nomad pid=3974
    2022-01-29T09:20:09.580Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: waiting for RPC address: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis path=/usr/bin/nomad
    2022-01-29T09:20:09.617Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon.nomad: plugin address: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis @module=logmon address=/tmp/plugin2044465918 network=unix timestamp=2022-01-29T09:20:09.617Z
    2022-01-29T09:20:09.618Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: using plugin: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis version=2
    2022-01-29T09:20:09.628Z [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis path=/tmp/NomadClient4275406614/070a3169-6c79-8cc5-32f5-4b80c61d6d10/alloc/logs/.redis.stdout.fifo @module=logmon timestamp=2022-01-29T09:20:09.627Z
    2022-01-29T09:20:09.628Z [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis @module=logmon path=/tmp/NomadClient4275406614/070a3169-6c79-8cc5-32f5-4b80c61d6d10/alloc/logs/.redis.stderr.fifo timestamp=2022-01-29T09:20:09.628Z
    2022-01-29T09:20:09.696Z [DEBUG] client.driver_mgr.nomad-driver-podman: Found imageID: driver=podman 4f97d722a1f0b864608aae4158d08b63d0d9ee301e6c3dec105ff57b21d5c350="for image" @module=podman docker.io/library/redis:3.2="in local storage" timestamp=2022-01-29T09:20:09.695Z
    2022-01-29T09:20:09.770Z [DEBUG] client: updated allocations: index=13 total=1 pulled=0 filtered=1
    2022-01-29T09:20:09.770Z [DEBUG] client: allocation updates: added=0 removed=0 updated=0 ignored=1
    2022-01-29T09:20:09.770Z [DEBUG] client: allocation updates applied: added=0 removed=0 updated=0 ignored=1 errors=0
    2022-01-29T09:20:10.572Z [DEBUG] http: request complete: method=GET path=/v1/evaluation/7c921fcf-641d-9a96-e2f1-7d955cc2fb72 duration="968.994µs"
    2022-01-29T09:20:10.582Z [DEBUG] http: request complete: method=GET path=/v1/evaluation/7c921fcf-641d-9a96-e2f1-7d955cc2fb72/allocations duration=1.964284ms
    2022-01-29T09:20:10.582Z [DEBUG] client.driver_mgr.nomad-driver-podman: Cleaning up: driver=podman @module=podman container=8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10 timestamp=2022-01-29T09:20:10.581Z
    2022-01-29T09:20:10.595Z [DEBUG] http: request complete: method=GET path="/v1/deployment/b724ae9c-c092-77dc-06d8-18576acccc8d?stale=&wait=2000ms" duration=2.787649ms
    2022-01-29T09:20:10.729Z [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis
  error=
  | rpc error: code = Unknown desc = failed to start task, could not start container: unknown error, status code: 500: {"cause":"exit status 1","message":"exit status 1","response":500}

    2022-01-29T09:20:10.729Z [INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10 task=redis reason="Error was unrecoverable"
    2022-01-29T09:20:10.730Z [INFO]  client.gc: marking allocation for GC: alloc_id=070a3169-6c79-8cc5-32f5-4b80c61d6d10
    2022-01-29T09:20:10.823Z [DEBUG] nomad.client: adding evaluations for rescheduling failed allocations: num_evals=1
    2022-01-29T09:20:10.828Z [DEBUG] client: updated allocations: index=14 total=1 pulled=0 filtered=1
    2022-01-29T09:20:10.828Z [DEBUG] http: request complete: method=GET path="/v1/deployment/b724ae9c-c092-77dc-06d8-18576acccc8d?index=11&stale=&wait=2000ms" duration=225.896419ms
    2022-01-29T09:20:10.829Z [DEBUG] client: allocation updates: added=0 removed=0 updated=0 ignored=1
    2022-01-29T09:20:10.829Z [DEBUG] client: allocation updates applied: added=0 removed=0 updated=0 ignored=1 errors=0
    2022-01-29T09:20:10.828Z [DEBUG] worker: dequeued evaluation: worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 eval_id=b99937cf-ac5a-b17f-2d8d-109aac1e39e5 type=service namespace=default job_id=redis node_id="" triggered_by=alloc-failure
    2022-01-29T09:20:10.830Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=b99937cf-ac5a-b17f-2d8d-109aac1e39e5 job_id=redis namespace=default worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3
  results=
  | Total changes: (place 0) (destructive 0) (inplace 0) (stop 0)
  | Desired Changes for "cache": (place 0) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 1) (canary 0)

    2022-01-29T09:20:10.831Z [DEBUG] worker.service_sched: setting eval status: eval_id=b99937cf-ac5a-b17f-2d8d-109aac1e39e5 job_id=redis namespace=default worker_id=a5f28f46-1dcb-d9d1-0ca8-da421bfe4fc3 status=complete
```

### Relevant part of the podman logs
```
DEBU[0208] Created OCI spec for container 8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10 at /var/lib/podman/.local/share/containers/storage/overlay-containers/8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10/userdata/config.json
DEBU[0208] /usr/libexec/podman/conmon messages will be logged to syslog
DEBU[0208] running conmon: /usr/libexec/podman/conmon    args="[--api-version 1 -c 8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10 -u 8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10 -r /usr/bin/crun -b /var/lib/podman/.local/share/containers/storage/overlay-containers/8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10/userdata -p /tmp/podman-run-978/containers/overlay-containers/8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10/userdata/pidfile -n redis-070a3169-6c79-8cc5-32f5-4b80c61d6d10 --exit-dir /tmp/podman-run-978/libpod/tmp/exits --full-attach -l k8s-file:/tmp/NomadClient4275406614/070a3169-6c79-8cc5-32f5-4b80c61d6d10/alloc/logs/.redis.stdout.fifo --log-level debug --syslog --conmon-pidfile /tmp/podman-run-978/containers/overlay-containers/8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10/userdata/conmon.pid --exit-command /usr/bin/podman --exit-command-arg --root --exit-command-arg /var/lib/podman/.local/share/containers/storage --exit-command-arg --runroot --exit-command-arg /tmp/podman-run-978/containers --exit-command-arg --log-level --exit-command-arg debug --exit-command-arg --cgroup-manager --exit-command-arg cgroupfs --exit-command-arg --tmpdir --exit-command-arg /tmp/podman-run-978/libpod/tmp --exit-command-arg --runtime --exit-command-arg crun --exit-command-arg --storage-driver --exit-command-arg overlay --exit-command-arg --storage-opt --exit-command-arg overlay.mount_program=/bin/fuse-overlayfs --exit-command-arg --events-backend --exit-command-arg file --exit-command-arg --syslog --exit-command-arg container --exit-command-arg cleanup --exit-command-arg 8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10]"
[conmon:e]: Failed to open log file Permission denied
DEBU[0208] Cleaning up container 8dcc56fa93316cea9dfd7c26cc635eb52f1999824ad143059913d64748c83c10
```

Hi,
I not sure to post answer to this issue. I think it's duplicate. Let me know if I need to open another issue.

Nomad 1.4.3
Nomad Podman Driver 0.4.1
Podman 4.3.1 (compil from source and rootless)
Ubuntu 22.04

nomad config podman.hcl
```
plugin "nomad-driver-podman" {
  config {
    socket_path = "unix:///run/user/1001/podman/podman.sock"
    volumes {
      enabled      = true
    }
  }
}
```



When I try the example podman job:

```
job "redis" {
  datacenters = ["dc1"]
  type        = "service"

  group "cache" {
    network {
      port "redis" { to = 6379 }
    }

    task "redis" {
      driver = "podman"

      config {
        image = "docker://redis"
        ports = ["redis"]
      }
    }
  }
}
```

Task failed:

```
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="/usr/local/bin/podman filtering at log level info"
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Setting parallel job count to 7"
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Using systemd socket activation to determine API endpoint"
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="API service listening on \"/run/user/1001/podman/podman.sock\". URI: \"/run/user/1001/podman/podman.sock\""
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Request Failed(Not Found): no container with name or ID \"redis-465b2ebb-0b8a-28f5-98a8-96395ac152ee\" found: no such container"
Jan  4 16:45:09 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:09 +0100] "GET /v1.0.0/libpod/containers/redis-465b2ebb-0b8a-28f5-98a8-96395ac152ee/json HTTP/1.1" 404 158 "" "Go-http-client/1.1"
Jan  4 16:45:09 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:09 +0100] "GET /v1.0.0/libpod/images/docker.io/library/redis:latest/json HTTP/1.1" 200 8205 "" "Go-http-client/1.1"
Jan  4 16:45:09 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:09 +0100] "POST /v1.0.0/libpod/containers/create HTTP/1.1" 201 88 "" "Go-http-client/1.1"
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Failed to add conmon to cgroupfs sandbox cgroup: creating cgroup path conmon: open /sys/fs/cgroup/cgroup.subtree_control: permission denied"
Jan  4 16:45:09 sandbox2 podman[58307]: [conmon:e]: Failed to open log file Permission denied
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Request Failed(Internal Server Error): exit status 1"
Jan  4 16:45:09 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:09 +0100] "POST /v1.0.0/libpod/containers/0f1fe7966aa0c593067bdfb4efc107847f05c0260c6c0908987320ddcf8b9e23/start HTTP/1.1" 500 67 "" "Go-http-client/1.1"
Jan  4 16:45:09 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:09 +0100] "DELETE /v1.0.0/libpod/containers/0f1fe7966aa0c593067bdfb4efc107847f05c0260c6c0908987320ddcf8b9e23?force=true&v=true HTTP/1.1" 200 154 "" "Go-http-client/1.1"
Jan  4 16:45:09 sandbox2 nomad[47651]:     2023-01-04T17:45:09.700+0100 [ERROR] client.driver_mgr.nomad-driver-podman: failed to clean up from an error in Start: driver=podman @module=podman error="cannot delete container, status code: 200" timestamp="2023-01-04T17:45:09.699+0100"
Jan  4 16:45:09 sandbox2 nomad[47651]: client.driver_mgr.nomad-driver-podman: failed to clean up from an error in Start: driver=podman @module=podman error="cannot delete container, status code: 200" timestamp="2023-01-04T17:45:09.699+0100"
Jan  4 16:45:09 sandbox2 nomad[47651]:     2023-01-04T17:45:09.701+0100 [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis
Jan  4 16:45:09 sandbox2 nomad[47651]:   error=
Jan  4 16:45:09 sandbox2 nomad[47651]:   | rpc error: code = Unknown desc = failed to start task, could not start container: cannot start container, status code: 500: {"cause":"exit status 1","message":"exit status 1","response":500}
Jan  4 16:45:09 sandbox2 nomad[47651]:   
Jan  4 16:45:09 sandbox2 nomad[47651]:     2023-01-04T17:45:09.701+0100 [INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis reason="Error was unrecoverable"
Jan  4 16:45:09 sandbox2 nomad[47651]: client.alloc_runner.task_runner: running driver failed: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis#012  error=#012  | rpc error: code = Unknown desc = failed to start task, could not start container: cannot start container, status code: 500: {"cause":"exit status 1","message":"exit status 1","response":500}#012  
Jan  4 16:45:09 sandbox2 nomad[47651]:  client.alloc_runner.task_runner: not restarting task: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis reason="Error was unrecoverable"
Jan  4 16:45:13 sandbox2 podman[58278]: @ - - [04/Jan/2023:17:45:13 +0100] "GET /libpod/_ping HTTP/1.1" 200 2 "" "Go-http-client/1.1"
Jan  4 16:45:13 sandbox2 nomad[47651]:     2023-01-04T17:45:13.707+0100 [WARN]  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis @module=logmon timestamp="2023-01-04T17:45:13.707+0100"
Jan  4 16:45:13 sandbox2 nomad[47651]:  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis @module=logmon timestamp="2023-01-04T17:45:13.707+0100"
Jan  4 16:45:13 sandbox2 nomad[47651]:     2023-01-04T17:45:13.708+0100 [WARN]  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis @module=logmon timestamp="2023-01-04T17:45:13.707+0100"
Jan  4 16:45:13 sandbox2 nomad[47651]:  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee task=redis @module=logmon timestamp="2023-01-04T17:45:13.707+0100"
Jan  4 16:45:13 sandbox2 nomad[47651]:     2023-01-04T17:45:13.711+0100 [INFO]  client.gc: marking allocation for GC: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee
Jan  4 16:45:13 sandbox2 nomad[47651]:  client.gc: marking allocation for GC: alloc_id=465b2ebb-0b8a-28f5-98a8-96395ac152ee
```

I can run any podman container, it works fine. `podman
As root or as podman rootless user.

```
root@sandbox2:/tmp/conmon# podman run -dt -p 8080:80 docker.io/nginx
e8c62f3814a2553a747ba30d1548304098402472835a4445250e7e957cc86144
root@sandbox2:/tmp/conmon# podman ps
CONTAINER ID  IMAGE                           COMMAND               CREATED        STATUS            PORTS                 NAMES
e8c62f3814a2  docker.io/library/nginx:latest  nginx -g daemon o...  5 seconds ago  Up 5 seconds ago  0.0.0.0:8080->80/tcp  awesome_northcutt
```

For me revelant logs:
```
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Failed to add conmon to cgroupfs sandbox cgroup: creating cgroup path conmon: open /sys/fs/cgroup/cgroup.subtree_control: permission denied"
Jan  4 16:45:09 sandbox2 podman[58307]: [conmon:e]: Failed to open log file Permission denied
Jan  4 16:45:09 sandbox2 podman[58278]: time="2023-01-04T17:45:09+01:00" level=info msg="Request Failed(Internal Server Error): exit status 1"
```

I cannot find solution to add conmon in  /sys/fs/cgroup/cgroup.subtree_control.

I don't know!

I also try Nomad this podman in root mode, it works!

Does someone already run Nomad with Podman in rootless ? 

Thanks!
And Good Year with Hashicorp!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,7 @@ IMPROVEMENTS:
 BUG FIXES:
 * api: Fixed file descriptor leak when getting logs from the task [[GH-508](https://github.com/hashicorp/nomad-driver-podman/pull/508)]
 * driver: Fixed static IP configuration (`static_ips`, `ipv4_address`, `ipv6_address`, `static_mac`) being silently ignored, corrected related per-network map handling and `driverNet.IP` resolution for service discovery. [[GH-507](https://github.com/hashicorp/nomad-driver-podman/pull/507)]
-* driver: Fixed log FIFO permission denied error when using rootless podman. [[GH-510](https://github.com/hashicorp/nomad-driver-podman/pull/510)]
+* driver: Fixed log FIFO permission denied error when using rootless podman with Nomad run as root. [[GH-510](https://github.com/hashicorp/nomad-driver-podman/pull/510)]
 
 ## 0.6.4 (December 10, 2025)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@ IMPROVEMENTS:
 BUG FIXES:
 * api: Fixed file descriptor leak when getting logs from the task [[GH-508](https://github.com/hashicorp/nomad-driver-podman/pull/508)]
 * driver: Fixed static IP configuration (`static_ips`, `ipv4_address`, `ipv6_address`, `static_mac`) being silently ignored, corrected related per-network map handling and `driverNet.IP` resolution for service discovery. [[GH-507](https://github.com/hashicorp/nomad-driver-podman/pull/507)]
+* driver: Fixed log FIFO permission denied error when using rootless podman. [[GH-510](https://github.com/hashicorp/nomad-driver-podman/pull/510)]
 
 ## 0.6.4 (December 10, 2025)
 

--- a/api/api.go
+++ b/api/api.go
@@ -1,4 +1,4 @@
-// Copyright IBM Corp. 2019, 2025
+// Copyright IBM Corp. 2019, 2026
 // SPDX-License-Identifier: MPL-2.0
 
 package api

--- a/api/api.go
+++ b/api/api.go
@@ -19,6 +19,7 @@ import (
 type API struct {
 	apiVersion       string
 	baseUrl          string
+	socketPath       string
 	defaultPodman    bool
 	appArmor         bool
 	cgroupV2         bool
@@ -54,6 +55,7 @@ func NewClient(logger hclog.Logger, config ClientConfig) *API {
 	ac := &API{
 		logger:        logger,
 		defaultPodman: config.DefaultPodman,
+		socketPath:    config.SocketPath,
 	}
 
 	ac.logger.Debug("http baseurl", "url", config.SocketPath)
@@ -107,6 +109,10 @@ func (c *API) SetRootless(isRootless bool) {
 
 func (c *API) IsRootless() bool {
 	return c.rootless
+}
+
+func (c *API) GetSocketPath() string {
+	return c.socketPath
 }
 
 func (c *API) SetAppArmor(appArmorEnabled bool) {

--- a/api/container_logs.go
+++ b/api/container_logs.go
@@ -1,4 +1,4 @@
-// Copyright IBM Corp. 2019, 2025
+// Copyright IBM Corp. 2019, 2026
 // SPDX-License-Identifier: MPL-2.0
 
 package api

--- a/driver.go
+++ b/driver.go
@@ -1965,12 +1965,15 @@ func resolveContainerIP(networkSettings *api.InspectNetworkSettings, networkName
 
 // ensureFifoAccessible ensures the log FIFO at fifoPath can be written to by
 // the rootless podman user. It determines the podman user by stat-ing the unix
-// socket file and chowns the FIFO to that user. Falls back to chmod 0666 if
-// the user cannot be determined. No-op when podman is running as root.
+// socket file and chowns the FIFO to that user. No-op when podman is running
+// as root or when the socket owner cannot be determined (e.g. TCP socket).
 func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *api.API) error {
+	// Nothing to do if log collection is disabled (no FIFO path configured).
 	if fifoPath == "" {
 		return nil
 	}
+	// Rootful podman: conmon runs as root and can already write to the
+	// root-owned FIFO, so no ownership change is needed.
 	if !podmanClient.IsRootless() {
 		return nil
 	}
@@ -1979,14 +1982,13 @@ func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *ap
 
 	if ok {
 		if err := os.Chown(fifoPath, uid, gid); err != nil {
-			logger.Warn("failed to chown fifo, falling back to chmod", "path", fifoPath, "error", err)
+			logger.Warn("failed to chown fifo", "path", fifoPath, "error", err)
 			return os.Chmod(fifoPath, 0666)
 		}
 	} else {
-		// Socket owner unknown (TCP or stat failed) — fallback to world-writable
-		if err := os.Chmod(fifoPath, 0666); err != nil {
-			return fmt.Errorf("failed to chmod fifo %s: %w", fifoPath, err)
-		}
+		// Cannot determine socket owner (TCP/HTTP socket or stat failed) — skip.
+		// Rootless podman over non-unix sockets is unsupported for log FIFOs.
+		logger.Debug("cannot determine podman socket owner, skipping fifo chown", "path", fifoPath)
 	}
 
 	return nil

--- a/driver.go
+++ b/driver.go
@@ -1003,13 +1003,17 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 	if !recoverRunningContainer {
 		// Ensure log FIFOs are accessible by the rootless podman user before starting
 		// the container. Nomad's logmon creates FIFOs owned by root:root with 0600;
-		// conmon (running as the podman user) needs write access.
+		// conmon (running as the podman user) needs write access. If chown fails,
+		// ContainerStart will also fail (conmon crashes with EACCES), so return
+		// early with a clearer error message.
 		if createOpts.LogConfiguration.Driver == "k8s-file" {
 			if fifoErr := ensureFifoAccessible(d.logger, cfg.StdoutPath, podmanClient); fifoErr != nil {
-				d.logger.Warn("failed to make stdout fifo accessible", "error", fifoErr)
+				cleanup()
+				return nil, nil, fmt.Errorf("failed to make stdout fifo accessible: %w", fifoErr)
 			}
 			if fifoErr := ensureFifoAccessible(d.logger, cfg.StderrPath, podmanClient); fifoErr != nil {
-				d.logger.Warn("failed to make stderr fifo accessible", "error", fifoErr)
+				cleanup()
+				return nil, nil, fmt.Errorf("failed to make stderr fifo accessible: %w", fifoErr)
 			}
 		}
 
@@ -1967,6 +1971,10 @@ func resolveContainerIP(networkSettings *api.InspectNetworkSettings, networkName
 // the rootless podman user. It determines the podman user by stat-ing the unix
 // socket file and chowns the FIFO to that user. No-op when podman is running
 // as root or when the socket owner cannot be determined (e.g. TCP socket).
+//
+// Uses os.Root to open the FIFO without following symlinks, preventing a
+// symlink swap attack where a restarted task replaces the FIFO with a symlink
+// to an arbitrary file.
 func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *api.API) error {
 	// Nothing to do if log collection is disabled (no FIFO path configured).
 	if fifoPath == "" {
@@ -1981,7 +1989,27 @@ func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *ap
 	uid, gid, ok := getSocketOwner(podmanClient.GetSocketPath())
 
 	if ok {
-		if err := os.Chown(fifoPath, uid, gid); err != nil {
+		// Open the FIFO's parent directory as a root to prevent symlink traversal.
+		dir := filepath.Dir(fifoPath)
+		base := filepath.Base(fifoPath)
+
+		root, err := os.OpenRoot(dir)
+		if err != nil {
+			return fmt.Errorf("failed to open fifo parent directory %q: %w", dir, err)
+		}
+		defer root.Close()
+
+		// Open the FIFO itself without following symlinks (os.Root enforces this).
+		// O_RDONLY|O_NONBLOCK avoids blocking on the FIFO (no writer needed).
+		// We only need a valid fd for fchown, not actual I/O.
+		f, err := root.OpenFile(base, os.O_RDONLY|syscall.O_NONBLOCK, 0)
+		if err != nil {
+			return fmt.Errorf("failed to open fifo %q: %w", fifoPath, err)
+		}
+		defer f.Close()
+
+		// Fchown on the file descriptor — safe from TOCTOU/symlink attacks.
+		if err := f.Chown(uid, gid); err != nil {
 			return err
 		}
 	} else {
@@ -1999,6 +2027,9 @@ func getSocketOwner(socketPath string) (int, int, bool) {
 	if !strings.HasPrefix(socketPath, "unix:") {
 		return 0, 0, false
 	}
+	// Two TrimPrefix calls handle both forms of the unix socket URI:
+	// "unix:/run/user/1001/podman/podman.sock"   → strip "unix:"
+	// "unix:///run/user/1001/podman/podman.sock"  → strip "unix:" then "//"
 	sockFile := strings.TrimPrefix(socketPath, "unix:")
 	sockFile = strings.TrimPrefix(sockFile, "//")
 

--- a/driver.go
+++ b/driver.go
@@ -1982,7 +1982,6 @@ func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *ap
 
 	if ok {
 		if err := os.Chown(fifoPath, uid, gid); err != nil {
-			logger.Warn("failed to chown fifo", "path", fifoPath, "error", err)
 			return err
 		}
 	} else {

--- a/driver.go
+++ b/driver.go
@@ -1005,11 +1005,11 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 		// the container. Nomad's logmon creates FIFOs owned by root:root with 0600;
 		// conmon (running as the podman user) needs write access.
 		if createOpts.LogConfiguration.Driver == "k8s-file" {
-			if err := ensureFifoAccessible(d.logger, cfg.StdoutPath, podmanClient); err != nil {
-				d.logger.Warn("failed to make stdout fifo accessible", "error", err)
+			if fifoErr := ensureFifoAccessible(d.logger, cfg.StdoutPath, podmanClient); fifoErr != nil {
+				d.logger.Warn("failed to make stdout fifo accessible", "error", fifoErr)
 			}
-			if err := ensureFifoAccessible(d.logger, cfg.StderrPath, podmanClient); err != nil {
-				d.logger.Warn("failed to make stderr fifo accessible", "error", err)
+			if fifoErr := ensureFifoAccessible(d.logger, cfg.StderrPath, podmanClient); fifoErr != nil {
+				d.logger.Warn("failed to make stderr fifo accessible", "error", fifoErr)
 			}
 		}
 

--- a/driver.go
+++ b/driver.go
@@ -14,6 +14,7 @@ import (
 	"slices"
 	"strconv"
 	"strings"
+	"syscall"
 	"time"
 
 	"github.com/armon/circbuf"
@@ -1000,6 +1001,18 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 	}
 
 	if !recoverRunningContainer {
+		// Ensure log FIFOs are accessible by the rootless podman user before starting
+		// the container. Nomad's logmon creates FIFOs owned by root:root with 0600;
+		// conmon (running as the podman user) needs write access.
+		if createOpts.LogConfiguration.Driver == "k8s-file" {
+			if err := ensureFifoAccessible(d.logger, cfg.StdoutPath, podmanClient); err != nil {
+				d.logger.Warn("failed to make stdout fifo accessible", "error", err)
+			}
+			if err := ensureFifoAccessible(d.logger, cfg.StderrPath, podmanClient); err != nil {
+				d.logger.Warn("failed to make stderr fifo accessible", "error", err)
+			}
+		}
+
 		if startErr := podmanClient.ContainerStart(d.ctx, containerID); startErr != nil {
 			cleanup()
 			return nil, nil, fmt.Errorf("failed to start task, could not start container: %w", startErr)
@@ -1948,4 +1961,53 @@ func resolveContainerIP(networkSettings *api.InspectNetworkSettings, networkName
 		return netData.IPAddress
 	}
 	return ""
+}
+
+// ensureFifoAccessible ensures the log FIFO at fifoPath can be written to by
+// the rootless podman user. It determines the podman user by stat-ing the unix
+// socket file and chowns the FIFO to that user. Falls back to chmod 0666 if
+// the user cannot be determined. No-op when podman is running as root.
+func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *api.API) error {
+	if fifoPath == "" {
+		return nil
+	}
+	if !podmanClient.IsRootless() {
+		return nil
+	}
+
+	uid, gid, ok := getSocketOwner(podmanClient.GetSocketPath())
+
+	if ok {
+		if err := os.Chown(fifoPath, uid, gid); err != nil {
+			logger.Warn("failed to chown fifo, falling back to chmod", "path", fifoPath, "error", err)
+			return os.Chmod(fifoPath, 0666)
+		}
+	} else {
+		// Socket owner unknown (TCP or stat failed) — fallback to world-writable
+		if err := os.Chmod(fifoPath, 0666); err != nil {
+			return fmt.Errorf("failed to chmod fifo %s: %w", fifoPath, err)
+		}
+	}
+
+	return nil
+}
+
+// getSocketOwner returns the UID/GID of a unix socket file.
+// Returns 0, 0, false if the owner cannot be determined.
+func getSocketOwner(socketPath string) (int, int, bool) {
+	if !strings.HasPrefix(socketPath, "unix:") {
+		return 0, 0, false
+	}
+	sockFile := strings.TrimPrefix(socketPath, "unix:")
+	sockFile = strings.TrimPrefix(sockFile, "//")
+
+	info, err := os.Stat(sockFile)
+	if err != nil {
+		return 0, 0, false
+	}
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, 0, false
+	}
+	return int(stat.Uid), int(stat.Gid), true
 }

--- a/driver.go
+++ b/driver.go
@@ -1983,7 +1983,7 @@ func ensureFifoAccessible(logger hclog.Logger, fifoPath string, podmanClient *ap
 	if ok {
 		if err := os.Chown(fifoPath, uid, gid); err != nil {
 			logger.Warn("failed to chown fifo", "path", fifoPath, "error", err)
-			return os.Chmod(fifoPath, 0666)
+			return err
 		}
 	} else {
 		// Cannot determine socket owner (TCP/HTTP socket or stat failed) — skip.

--- a/driver_test.go
+++ b/driver_test.go
@@ -3089,8 +3089,6 @@ func TestEnsureFifoAccessible(t *testing.T) {
 		expectErr  string
 		// If set, verify fifo ownership matches socket owner
 		verifyChown bool
-		// If set, verify fifo permissions are 0666
-		verifyChmod bool
 		// If true, skip creating a FIFO (for empty/noop cases)
 		skipFifo bool
 		// If true, test requires root
@@ -3172,12 +3170,6 @@ func TestEnsureFifoAccessible(t *testing.T) {
 				fifoStat := info.Sys().(*syscall.Stat_t)
 				must.Eq(t, int(sockStat.Uid), int(fifoStat.Uid), must.Sprint("fifo uid should match socket owner after chown"))
 				must.Eq(t, int(sockStat.Gid), int(fifoStat.Gid), must.Sprint("fifo gid should match socket owner after chown"))
-			}
-
-			if tc.verifyChmod {
-				info, statErr := os.Stat(fifoPath)
-				must.NoError(t, statErr)
-				must.Eq(t, os.FileMode(0666)|os.ModeNamedPipe, info.Mode(), must.Sprint("fifo should be world-writable (0666) after chmod fallback"))
 			}
 		})
 	}

--- a/driver_test.go
+++ b/driver_test.go
@@ -20,6 +20,7 @@ import (
 	"runtime"
 	"strconv"
 	"strings"
+	"syscall"
 	"testing"
 	"time"
 
@@ -2988,4 +2989,266 @@ func TestResolveContainerIP(t *testing.T) {
 			must.Eq(t, tc.expectedIP, result)
 		})
 	}
+}
+
+// TestGetSocketOwner verifies getSocketOwner returns the correct uid/gid
+// for unix sockets and false for non-unix or missing paths.
+func TestGetSocketOwner(t *testing.T) {
+	ci.Parallel(t)
+
+	// Use /tmp to avoid macOS socket path length limit (108 chars)
+	dir, err := os.MkdirTemp("/tmp", "sock-test-*")
+	must.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	sockPath := filepath.Join(dir, "t.sock")
+	listener, err := net.Listen("unix", sockPath)
+	must.NoError(t, err)
+	defer listener.Close()
+
+	// Get the actual socket ownership for assertions
+	sockInfo, err := os.Stat(sockPath)
+	must.NoError(t, err)
+	sockStat := sockInfo.Sys().(*syscall.Stat_t)
+
+	testCases := []struct {
+		name      string
+		input     string
+		expectOk  bool
+		expectUid int
+		expectGid int
+	}{
+		{
+			name:      "unix colon prefix",
+			input:     "unix:" + sockPath,
+			expectOk:  true,
+			expectUid: int(sockStat.Uid),
+			expectGid: int(sockStat.Gid),
+		},
+		{
+			name:      "unix double slash prefix",
+			input:     "unix://" + sockPath,
+			expectOk:  true,
+			expectUid: int(sockStat.Uid),
+			expectGid: int(sockStat.Gid),
+		},
+		{
+			name:  "http scheme returns false",
+			input: "http://localhost:8080",
+		},
+		{
+			name:  "tcp scheme returns false",
+			input: "tcp://127.0.0.1:8080",
+		},
+		{
+			name:  "empty string returns false",
+			input: "",
+		},
+		{
+			name:  "non-existent socket returns false",
+			input: "unix:///no/such/path/test.sock",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			uid, gid, ok := getSocketOwner(tc.input)
+			must.Eq(t, tc.expectOk, ok, must.Sprintf("getSocketOwner(%q) ok mismatch", tc.input))
+			if tc.expectOk {
+				must.Eq(t, tc.expectUid, uid, must.Sprintf("getSocketOwner(%q) uid mismatch", tc.input))
+				must.Eq(t, tc.expectGid, gid, must.Sprintf("getSocketOwner(%q) gid mismatch", tc.input))
+			}
+		})
+	}
+}
+
+// TestEnsureFifoAccessible verifies the FIFO accessibility logic for rootless podman.
+func TestEnsureFifoAccessible(t *testing.T) {
+	ci.Parallel(t)
+
+	// Use /tmp to avoid macOS socket path length limit (108 chars)
+	dir, err := os.MkdirTemp("/tmp", "fifo-test-*")
+	must.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	// Create a unix socket for tests that need it
+	sockPath := filepath.Join(dir, "t.sock")
+	listener, err := net.Listen("unix", sockPath)
+	must.NoError(t, err)
+	defer listener.Close()
+
+	sockInfo, err := os.Stat(sockPath)
+	must.NoError(t, err)
+	sockStat := sockInfo.Sys().(*syscall.Stat_t)
+
+	testCases := []struct {
+		name       string
+		fifoPath   string
+		rootless   bool
+		socketPath string
+		expectErr  string
+		// If set, verify fifo ownership matches socket owner
+		verifyChown bool
+		// If set, verify fifo permissions are 0666
+		verifyChmod bool
+		// If true, skip creating a FIFO (for empty/noop cases)
+		skipFifo bool
+		// If true, test requires root
+		requireRoot bool
+	}{
+		{
+			name:     "empty path is no-op",
+			fifoPath: "",
+			rootless: true,
+			skipFifo: true,
+		},
+		{
+			name:       "rootful is no-op even with bad path",
+			fifoPath:   "/no/such/fifo",
+			rootless:   false,
+			socketPath: "unix:" + sockPath,
+			skipFifo:   true,
+		},
+		{
+			name:        "chown to socket owner",
+			rootless:    true,
+			socketPath:  "unix:" + sockPath,
+			verifyChown: true,
+		},
+		{
+			name:        "chmod fallback for tcp socket path",
+			rootless:    true,
+			socketPath:  "http://localhost:8080",
+			verifyChmod: true,
+		},
+		{
+			name:       "non-existent fifo returns error",
+			fifoPath:   "/no/such/fifo/path",
+			rootless:   true,
+			socketPath: "http://localhost:8080",
+			expectErr:  "failed to chmod fifo",
+			skipFifo:   true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.requireRoot && os.Getuid() != 0 {
+				t.Skip("test requires root")
+			}
+
+			logger := testlog.HCLogger(t)
+
+			fifoPath := tc.fifoPath
+			if !tc.skipFifo {
+				fifoPath = filepath.Join(dir, fmt.Sprintf("fifo-%s", tc.name))
+				must.NoError(t, exec.Command("mkfifo", fifoPath).Run())
+				must.NoError(t, os.Chmod(fifoPath, 0600))
+			}
+
+			var client *api.API
+			if tc.socketPath != "" {
+				client = api.NewClient(logger, api.ClientConfig{
+					SocketPath:  tc.socketPath,
+					HttpTimeout: 5 * time.Second,
+				})
+			} else {
+				client = &api.API{}
+			}
+			client.SetRootless(tc.rootless)
+
+			err := ensureFifoAccessible(logger, fifoPath, client)
+
+			if tc.expectErr != "" {
+				must.Error(t, err, must.Sprint("ensureFifoAccessible should have returned an error"))
+				must.ErrorContains(t, err, tc.expectErr)
+				return
+			}
+			must.NoError(t, err, must.Sprint("ensureFifoAccessible should not have returned an error"))
+
+			if tc.verifyChown {
+				info, statErr := os.Stat(fifoPath)
+				must.NoError(t, statErr)
+				fifoStat := info.Sys().(*syscall.Stat_t)
+				must.Eq(t, int(sockStat.Uid), int(fifoStat.Uid), must.Sprint("fifo uid should match socket owner after chown"))
+				must.Eq(t, int(sockStat.Gid), int(fifoStat.Gid), must.Sprint("fifo gid should match socket owner after chown"))
+			}
+
+			if tc.verifyChmod {
+				info, statErr := os.Stat(fifoPath)
+				must.NoError(t, statErr)
+				must.Eq(t, os.FileMode(0666)|os.ModeNamedPipe, info.Mode(), must.Sprint("fifo should be world-writable (0666) after chmod fallback"))
+			}
+		})
+	}
+}
+
+// TestPodmanDriver_LogFifoAccessible is an integration test that verifies
+// the k8s-file log driver path works end-to-end: the driver makes FIFOs
+// accessible to rootless podman before starting the container, and container
+// output is captured correctly through the FIFO.
+func TestPodmanDriver_LogFifoAccessible(t *testing.T) {
+	ci.Parallel(t)
+
+	stdoutMagic := uuid.Generate()
+	stderrMagic := uuid.Generate()
+
+	taskCfg := newTaskConfig("", []string{
+		"sh",
+		"-c",
+		fmt.Sprintf("echo %s; 1>&2 echo %s", stdoutMagic, stderrMagic),
+	})
+	// Use default logging (maps to k8s-file internally), which is the path
+	// that triggers ensureFifoAccessible before container start.
+	taskCfg.Logging.Driver = "nomad"
+
+	task := &drivers.TaskConfig{
+		ID:        uuid.Generate(),
+		Name:      "logFifoAccess",
+		AllocID:   uuid.Generate(),
+		Resources: createBasicResources(),
+	}
+	must.NoError(t, task.EncodeConcreteDriverConfig(&taskCfg))
+
+	d := podmanDriverHarness(t, nil)
+	cleanup := d.MkAllocDir(task, true)
+	defer cleanup()
+
+	// Verify the stdout FIFO exists before start
+	_, err := os.Stat(task.StdoutPath)
+	must.NoError(t, err)
+
+	_, _, err = d.StartTask(task)
+	must.NoError(t, err)
+
+	defer func() {
+		_ = d.DestroyTask(task.ID, true)
+	}()
+
+	// After StartTask, verify the FIFO permissions were made accessible.
+	// In rootless mode this will be chowned; in rootful mode it's unchanged.
+	// Either way, the FIFO should still exist and be writable.
+	fifoInfo, err := os.Stat(task.StdoutPath)
+	must.NoError(t, err)
+	fifoMode := fifoInfo.Mode()
+	// The FIFO should be a named pipe
+	must.True(t, fifoMode&os.ModeNamedPipe != 0, must.Sprint("stdout fifo should be a named pipe after StartTask"))
+
+	// Wait for container to finish
+	waitCh, err := d.WaitTask(context.Background(), task.ID)
+	must.NoError(t, err, must.Sprint("WaitTask should not fail"))
+
+	select {
+	case res := <-waitCh:
+		must.Eq(t, 0, res.ExitCode, must.Sprint("container should exit cleanly"))
+	case <-time.After(10 * time.Second):
+		t.Fatalf("Container did not exit in time")
+	}
+
+	// Verify logs were captured through the FIFO — this confirms conmon
+	// could write to the FIFO (the original bug would cause empty logs
+	// or a permission denied startup failure).
+	stdoutLog := readStdoutLog(t, task)
+	must.StrContains(t, stdoutLog, stdoutMagic)
+	must.StrContains(t, stdoutLog, stderrMagic)
 }

--- a/driver_test.go
+++ b/driver_test.go
@@ -3091,6 +3091,8 @@ func TestEnsureFifoAccessible(t *testing.T) {
 		verifyChown bool
 		// If true, skip creating a FIFO (for empty/noop cases)
 		skipFifo bool
+		// If true, create a symlink instead of a FIFO (for symlink attack test)
+		createSymlink bool
 		// If true, test requires root
 		requireRoot bool
 	}{
@@ -3127,6 +3129,14 @@ func TestEnsureFifoAccessible(t *testing.T) {
 			skipFifo:   true,
 			// Should not error — just logs and returns nil
 		},
+		{
+			name:          "rejects symlink (TOCTOU protection)",
+			rootless:      true,
+			socketPath:    "unix:" + sockPath,
+			expectErr:     "failed to open fifo",
+			createSymlink: true,
+			skipFifo:      true,
+		},
 	}
 
 	for _, tc := range testCases {
@@ -3142,6 +3152,14 @@ func TestEnsureFifoAccessible(t *testing.T) {
 				fifoPath = filepath.Join(dir, fmt.Sprintf("fifo-%s", tc.name))
 				must.NoError(t, exec.Command("mkfifo", fifoPath).Run())
 				must.NoError(t, os.Chmod(fifoPath, 0600))
+			}
+			if tc.createSymlink {
+				// Simulate a symlink swap attack: task replaces FIFO with a
+				// symlink pointing to an arbitrary file (e.g. /etc/shadow).
+				fifoPath = filepath.Join(dir, fmt.Sprintf("fifo-%s", tc.name))
+				target := filepath.Join(dir, "attack-target")
+				must.NoError(t, os.WriteFile(target, []byte("sensitive"), 0600))
+				must.NoError(t, os.Symlink(target, fifoPath))
 			}
 
 			var client *api.API

--- a/driver_test.go
+++ b/driver_test.go
@@ -3262,6 +3262,7 @@ func TestPodmanDriver_LogFifoAccessible(t *testing.T) {
 	must.StrContains(t, stdoutLog, stdoutMagic)
 	must.StrContains(t, stdoutLog, stderrMagic)
 }
+
 // TestPodmanDriver_CustomNetwork validates custom network_mode behavior:
 // successful attachment, non-existent network error, and unreachable socket error.
 func TestPodmanDriver_CustomNetwork(t *testing.T) {

--- a/driver_test.go
+++ b/driver_test.go
@@ -3116,18 +3116,18 @@ func TestEnsureFifoAccessible(t *testing.T) {
 			verifyChown: true,
 		},
 		{
-			name:        "chmod fallback for tcp socket path",
-			rootless:    true,
-			socketPath:  "http://localhost:8080",
-			verifyChmod: true,
+			name:       "noop for tcp socket path",
+			rootless:   true,
+			socketPath: "http://localhost:8080",
+			// TCP socket — cannot determine owner, should noop (no chown, no chmod)
 		},
 		{
-			name:       "non-existent fifo returns error",
+			name:       "noop for non-existent fifo with tcp socket",
 			fifoPath:   "/no/such/fifo/path",
 			rootless:   true,
 			socketPath: "http://localhost:8080",
-			expectErr:  "failed to chmod fifo",
 			skipFifo:   true,
+			// Should not error — just logs and returns nil
 		},
 	}
 

